### DISCUSSION TITLE
Help understanding pcomplete and completion metadata

### DISCUSSION THREAD

**Discussion by karthink** [minad/corfu#419] [General]
Hi @minad,

I'm trying to improve Org mode's (rather abysmal) completion support, but I'm having some trouble understanding how `pcomplete` works.  I thought you might be the best person to ask for help/clarifications here.

Here's an example of what I have so far:

https://github.com/minad/corfu/assets/8607532/d8e2293b-e91f-469b-a691-58c4763f1e00

### Trouble with annotations

Annotations disappear once I start typing, sometimes they reappear again:

https://github.com/minad/corfu/assets/8607532/3b929c4d-23da-462b-b5fa-b09316c4c778

For some background, Org uses `pcomplete-completions-at-point` as its capf.  This function looks for an annotation attached to a candidate as a `pcomplete-annotation` text property, which I provide.

Any idea what might be causing this?

### What is `:company-docsig`?

`pcomplete-completions-at-point` looks for a `pcomplete-help` candidate text property and uses it to provide a `:company-docsig` key in the completion table metadata it generates.  What is `:company-docsig`?  When I added the `pcomplete-help` property to the candidates, I couldn't access it when using Corfu or the default completion-in-region-function.

### Showing documentation

I want to provide additional information/documentation about the completion when pressing `M-h` in Corfu.  For example, In the above example I would like to pop up a buffer showing an example of a bibliography entry in the chosen style.  From `corfu-info-documentation` it looks like this is through the use of a `:company-doc-buffer` key in the completion metadata.  However `pcomplete-completion-at-point` doesn't attach this key to the table it generates.  

1. Any idea how I can specify a function as the value of the `:company-doc-buffer` key when using pcomplete?
2. Is this feature accessible when using Emacs' default in-buffer completion function?

EDIT: The code I used to provide completions for `#+cite_export` is [here](https://gist.github.com/karthink/4b6f1d37672b704841b2c5fc944d9c73).

**Comment by minad** [comment]
Hi @karthink,

I think we can shorten the discussion - Org should move away from pcomplete and provide Capfs instead. :)

The additional pcomplete layer doesn't help. As far as I know there have been bugs where the metadata got lost in the pcomplete layer (fixed in 29 or 30).

> What is :company-docsig?

See `corfu-echo-mode`. You should find this when you grep the Corfu source.

> Showing documentation

Yes, `:company-doc-buffer` can be used to provide this. This feature is not accessible in default completion as far as I am aware.

**Comment by karthink** [comment]
> I think we can shorten the discussion - Org should move away from pcomplete and provide Capfs instead. :)

I wanted to add to `org-pcomplete.el` since it provides a bunch of the infrastructure already.  For example, I was able to write just the function `pcomplete/org-mode/file-option/cite_export` to complete the `cite_export` keyword and not worry about writing the logic to identify where the cursor is as a capf would.  How difficult do you think it is to write a capf that can handle the Org syntax and provide the context required for completions like the above? Each argument in the chain

```
#+cite_export: basic numeric author-year
```
uses a different completion table, and the pcomplete layer provides handy variables (`pcomplete-index`, `pcomplete-last` etc) to identify the cursor position.  I'm not sure how to do this with a capf.

Ultimately I don't care about the pcomplete/capf issue: I just want much better completion when writing in Org.  It's annoying to keep having to switch to the manual for information that could readily be provided via a completion annotation or the `:company-doc-buffer` system.

Thanks for the pointer to `corfu-echo-mode`, I'll check it out.

**Reply by minad** [reply]
You may be right that pcomplete already provides the right infrastructure for dispatching. It would certainly require a little work to create a similar mechanism. The Org syntax handling is already there in `org-thing-at-point`, all that is used from pcomplete is the dispatcher it seems.

From my experience the pcomplete mechanism has been broken for quite some time and lacks flexibility. It only got better in 29 and 30. The pcomplete mechanism was retrofitted onto the Capf mechanism at some point. 

pcomplete lacks the metadata `:company-doc-buffer`. You could use an advice to add it, see `pcomplete-completions-at-point` or propose to add this metadata upstream. Then one could attach the metadata as text properties to the candidates, but this won't be ideal. It would be better if each of the completion functions could return their own doc buffer function (instead of attaching a text property), but at this point things get hairy and it might be easier to avoid the pcomplete layer.

**Reply by karthink** [reply]
> The Org syntax handling is already there in org-thing-at-point, all that is used from pcomplete is the dispatcher it seems.

You're right, I won't be starting from scratch.  I'll take a crack at converting `org-pcomplete.el` from pcomplete to a capf, and go from there.
